rfcs/rfc-metaprogramming.md

```diff
@@ -0,0 +1,379 @@
+# Metaprogramming API
+
+## Preamble
+
+    Author:  Paul Evans <PEVANS>
+    Sponsor:
+    ID:      TODO
+    Status:  Exploratory
+
+## Abstract
+
+A coherent API for metaprogramming tasks; that is, code which is aware of its own code structure and can inspect or manipulate it.
+
+## Motivation
+
+As a dynamic language, programs written in Perl have the ability to be aware of the structure of their own code. Functions, packages, variables, and other constructions can be obtained or manipulated by name, While the techniques can certainly be over-used, it is a useful and necessary ability of what it means to be a "dynamic language", and where appropriate these kinds of techniques should be supported and encouraged.
+
+Originally, Perl provided some abilities in the form of operations on glob references. Gradually over time, authors have required more abilities than this and increasingly have used a variety of CPAN modules to extend these abilities. As many of the operations provided ought to be considered fundamental language abilities for a dynamic language, it would be better if these were provided by core Perl itself.
+
+This RFC aims to provide a set of API functions to cover many of these abilities. Similar to the case of RFC 0009 having added a namespace for general "builtin" functions, it is not intended that this RFC covers a total and complete list; but instead sets out an initial set of functions. It is expected that more would be added over time on an individual basis, as the need arises. In particular, this RFC does not address any additional API that may be needed by the new object system currently being developed under the `use feature 'class'` flag.
+
+## Rationale
+
+(explain why the (following) proposed solution will solve it)
+
+## Specification
+
+As this RFC adds a number of functions all on an overall-similar theme, they are all collected into a new toplevel namespace. For now this namespace is called `meta::`; though as per the rest of the details in this RFC that is currently very much up for debate.
+
+Objects that are returned by these functions will be instances of some named package within the `meta::` namespace, though we do not yet specify exactly what they will be. Furthermore we make no statement about what actual container type of objects those actually are. Users should not attempt to interact with these objects except via the API specified here.
+
+### Functions to operate on Packages
+
+These are functions to create, manipulate, and inspect packages themselves.
+
+#### `has_package`
+
+```perl
+$bool = meta::has_package($name);
+```
+
+Returns true if the named package exists.
```

Comment:
What does exits mean in this context?
Is it loaded by the current interpreter?
Installed and available in in a path on PERL5LIB?